Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,365 @@
+        private const string ExceptionEventName = "System.Net.Http.Exception";
+        private const string InitializationFailed = "System.Net.Http.InitializationFailed";
+
+        // Fields for reflection
+        private static FieldInfo connectionGroupListField;
+        private static Type connectionGroupType;
+        private static FieldInfo connectionListField;
+        private static Type connectionType;
+        private static FieldInfo writeListField;
+        private static Func<HttpWebRequest, HttpWebResponse> httpResponseAccessor;
+
+        // Fields for controlling initialization of the HttpHandlerDiagnosticListener singleton
+        private object initializationLock = new object();
+        private bool initialized = false;
+#endregion
+
+        /// <summary>
+        /// Prevents a default instance of the <see cref="HttpHandlerDiagnosticListener" /> class from being created.
+        /// This class implements a singleton pattern and only this class is allowed to create an instance.
+        /// </summary>
+        private HttpHandlerDiagnosticListener() : base(DiagnosticListenerName)
+        {
+        }
+
+        /// <summary>
+        /// Overriding base class implementation just to give us a chance to initialize.  
+        /// </summary>
+        public override IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Predicate<string> isEnabled)
+        {
+            this.Initialize();
+            return base.Subscribe(observer, isEnabled);
+        }
+
+        /// <summary>
+        /// Overriding base class implementation just to give us a chance to initialize.  
+        /// </summary>
+        public override IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Func<string, object, object, bool> isEnabled)
+        {
+            this.Initialize();
+            return base.Subscribe(observer, isEnabled);
+        }
+
+#region private methods
+
+        private static void PrepareReflectionObjects()
+        {
+            // At any point, if the operation failed, it should just throw, including NullReferenceException.
+            // The caller should catch all exceptions and swallow.
+
+            // First step: Get all the reflection objects we will ever need.
+            Assembly systemNetHttpAssembly = typeof(ServicePoint).Assembly;
+            connectionGroupListField = typeof(ServicePoint).GetField("m_ConnectionGroupList", BindingFlags.Instance | BindingFlags.NonPublic);
+            connectionGroupType = systemNetHttpAssembly?.GetType("System.Net.ConnectionGroup");
+            connectionListField = connectionGroupType?.GetField("m_ConnectionList", BindingFlags.Instance | BindingFlags.NonPublic);
+            connectionType = systemNetHttpAssembly?.GetType("System.Net.Connection");
+            writeListField = connectionType?.GetField("m_WriteList", BindingFlags.Instance | BindingFlags.NonPublic);
```

Comment:
It's better to just attempt the operation rather than checking it. If you have no access, it will just fail, and execution goes on. It falls back to just framework events without diagnostic source and no harm is done.